auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java

```diff
@@ -154,17 +155,28 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
     public void onSuccess(final IDPResponse response) {
         AuthCredential credential = createCredential(response);
         final FirebaseAuth firebaseAuth = mActivityHelper.getFirebaseAuth();
-
-        firebaseAuth
-                .signInWithCredential(credential)
-                .addOnFailureListener(
-                        new TaskFailureLogger(TAG, "Firebase sign in with credential unsuccessful"))
-                .addOnCompleteListener(new CredentialSignInHandler(
-                        AuthMethodPickerActivity.this,
-                        mActivityHelper,
-                        RC_ACCOUNT_LINK,
-                        RC_SAVE_CREDENTIAL,
-                        response));
+        FirebaseUser user = firebaseAuth.getCurrentUser();
+
+        TaskFailureLogger logger = new TaskFailureLogger(
+                TAG, "Firebase sign in with credential unsuccessful");
+        CredentialSignInHandler handler = new CredentialSignInHandler(
+                AuthMethodPickerActivity.this,
+                mActivityHelper,
+                RC_ACCOUNT_LINK,
+                RC_SAVE_CREDENTIAL,
+                response);
+
+        if (user != null && mActivityHelper.getFlowParams().shouldLinkUser) {
+            user
+                    .linkWithCredential(credential)
```

Comment:
style nit: prefer not to have a continuation for the first method call in a chain if it would fit on the previous line.  `user` is just kinda floating out there.